### PR TITLE
refactor: AST rework

### PR DESCRIPTION
**Pull request by elenakrittik**

### CODE REVIEW

**Comment by elenakrittik** [comment]
* **#46** <a href="https://app.graphite.dev/github/pr/metallang/metal/46?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @elenakrittik and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/metal-ast-ng/metal.ungram [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+StructBody =
+  items:(StructBodyItem (',' StructBodyItem)* ','?)
+
+StructBodyItem =
+  StructField | FnItem
+
+StructField =
+  vis:Visibility Name ty:TypeQualifier
+
+Type =
```
Comment (VincentRPS) [inline crates/metal-ast-ng/metal.ungram:82]: It seems this does no AST interpretation of the inner types? (Vec[T])

**Review by VincentRPS — CHANGES_REQUESTED** [review]

File: crates/metal-ast-ng/metal.ungram [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+PrefixExpr =
+  op:PrefixOp Expr
+
+PrefixOp =
+  '+' | '-' | '!' | '~'
+
+BinaryExpr =
+  lhs:Expr op:BinaryOp rhs:Expr
+
+BinaryOp =
```
[the remark this answers is not in the document]
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram:97]: You mean overflows?

File: crates/metal-ast-ng/metal.ungram (hunk 1 above)
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram:82]: Generics are indeed not in the grammar

File: crates/metal-ast-ng/metal.ungram [hunk 3]
```diff
@@ -0,0 +1,153 @@
+Name =
```
[the remark this answers is not in the document]
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram:1]: I assume abstracts can only contain bodyless functions? Also, references can be `&T` and `&mut T`, correct?

[the remark this answers is not in the document]
Comment (VincentRPS) [reply crates/metal-ast-ng/metal.ungram:1]: > I assume abstracts can only contain bodyless functions? Also, references can be `&T` and `&mut T`, correct?

correct

File: crates/metal-ast-ng/metal.ungram (hunk 2 above)
[the remark this answers is not in the document]
Comment (VincentRPS) [reply crates/metal-ast-ng/metal.ungram:97]: > You mean overflows?

yes

[the remark this answers is not in the document]
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram:97]: I don't think overflows are a good reason to remove operators. Not only operators are overloadable, i.e. overflows are not a concern with every usage of `+`/etc., but overflows are also rarely a big concern in code. For cases where overflows are undesirable, an off-by-default lint can be trivially implemented that will check for potential overflows (or even require marking them as unsafe or something).

[the remark this answers is not in the document]
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram:97]: Also, with the exception of `**` and `**=`, Rust has all of these operators

File: crates/metal-ast-ng/metal.ungram (hunk 1 above)
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram]: Done!

File: crates/metal-ast-ng/metal.ungram [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  '{' subtrees:ImportBranchSubtrees? '}'
+
+ImportBranchSubtrees =
+  ImportTree (',' ImportTree)* ','?
+
+StructItem =
+  vis:Visibility 'struct' Name '{' body:StructBody '}'
+
+StructBody =
+  items:(StructBodyItem (',' StructBodyItem)* ','?)
```
[the remark this answers is not in the document]
Comment (elenakrittik) [reply crates/metal-ast-ng/metal.ungram]: Done!

**Review by VincentRPS — APPROVED** [review]
LGTM!

**Pull request merged**